daemons/ipa-kdb/ipa_kdb_kdcpolicy.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (C) 2018  FreeIPA Contributors see COPYING for license
+ */
+
+#include <errno.h>
+#include <syslog.h>
+#include <krb5/kdcpolicy_plugin.h>
+
+#include "ipa_krb5.h"
+#include "ipa_kdb.h"
+
+struct krb5_kdcpolicy_moddata_st {
+    /* Storage struct for kdcpolicy callback state */
+    ;
+};
+
+static krb5_error_code ipa_kdcpolicy_init(krb5_context context,
+                                          krb5_kdcpolicy_moddata *data_out)
+{
+    struct krb5_kdcpolicy_moddata_st *data = calloc(1, sizeof(*data));
+    if (data == NULL) {
+        return ENOMEM;
+    }
+
+    krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy plugin loaded.");
+    *data_out = data;
+
+    return 0;
+}
+
+static krb5_error_code ipa_kdcpolicy_fini(krb5_context context,
+                                          krb5_kdcpolicy_moddata moddata)
+{
+    free(moddata);
+    krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy plugin unloaded.");
+    return 0;
+}
+
+static krb5_error_code
+ipa_kdcpolicy_check_as(krb5_context context, krb5_kdcpolicy_moddata moddata,
+                       const krb5_kdc_req *request,
+                       const krb5_db_entry *client,
+                       const krb5_db_entry *server,
+                       const char *const *auth_indicators,
+                       const char **status, krb5_deltat *lifetime_out,
+                       krb5_deltat *renew_lifetime_out)
+{
+    krb5_error_code kerr;
+    enum ipadb_user_auth ua;
+    struct ipadb_e_data *ied;
+    int valid_auth_indicators = 0;
+
+    *status = NULL;
+    *lifetime_out = 0;
+    *renew_lifetime_out = 0;
+
+    krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: checking AS-REQ.");
+
+    ied = (struct ipadb_e_data *)client->e_data;
+    if (ied == NULL || ied->magic != IPA_E_DATA_MAGIC) {
+        /* e-data is not availble, getting user auth from LDAP */
+        krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: client e_data not availble. Try fetching...");
+        kerr = ipadb_get_principal(context, request->client, KRB5_KDB_FLAG_ALIAS_OK, &client);
+        if (kerr != 0) {
+            krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: ipadb_find_principal failed.");
+            return kerr;
+        }
+
+        ied = (struct ipadb_e_data *)client->e_data;
+        if (ied == NULL && ied->magic != IPA_E_DATA_MAGIC) {
+            krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: client e_data fetching failed.");
+            return EINVAL;
+        }
+    }
+
+    ua = ied->user_auth;
+    
+    /* If no mechanisms are set, allow every auth method */
+    if (ua == IPADB_USER_AUTH_NONE) {
+        return 0;
+    }
+
+    /* For each auth indicator, see if it is allowed for that user */
+    for (int i = 0; auth_indicators[i] != NULL; i++) {
+        const char *auth_indicator = auth_indicators[i];
+
+        if (strcmp(*auth_indicators, "otp") == 0) {
```

Comment:
*auth_indicators -> auth_indicator and everywhere below

Reply:
I know why this worked before, but I'm still surprised that it actually did...